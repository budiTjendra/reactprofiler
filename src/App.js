import React, {useState} from 'react';
import './App.css';
import {FixedSizeList as List} from "react-window";
import AutoSizer from "react-virtualized-auto-sizer";
import * as api from "./utils/api";

import "./style.css";

function App() {
    const [data, setData] = useState(undefined)

    const loadData = async () => {
        const colorData = await api.fetchData();
        setData(colorData)
    };

    const dataSize = (() => {
        if (data === undefined)
            return 0

        return Object.keys(data).length
    })();

    const Row = (props) => {
        const {
            index, style
        } = props
        const colorName = data !== undefined ? Object.keys(data)[index] : undefined;
        const colorStyle = data !== undefined ?
            (() => {
                const rgbVal = Object.values(data)[index]
                return ({backgroundColor: `rgba(${rgbVal[0]},${rgbVal[1]},${rgbVal[2]},${rgbVal[3]})`, ...style})
                //return (style)
            })() : style;
        return (
            <div
                className={index % 2 ? "ListItemOdd" : "ListItemEven"}
                style={colorStyle}>
                {colorName}
            </div>
        );
    }


    const Example = () => {
        return (

            <>
                {
                    data && (<div role="output" data-testid="output">data loaded!</div>)
                }
                <AutoSizer>
                    {({height, width}) => (
                        <List
                            role={'List'}
                            className="List"
                            height={height}
                            itemCount={dataSize}
                            itemSize={35}
                            width={width}
                        >
                            {Row}
                        </List>
                    )}
                </AutoSizer>
            </>
        );

    }

    return (
        <div className="App-fullscreen" >
            <button data-testid="show-color-button" onClick={loadData}>show color</button>

            <Example/>


        </div>
    );
}

export default App;
