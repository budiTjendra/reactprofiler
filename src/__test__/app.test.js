import React from 'react';
import renderer from 'react-test-renderer';
import App from '../App';
import { render, fireEvent, waitForElement } from '@testing-library/react'

test('render correctly', () =>{
    const tree = renderer
        .create(<App/>)
        .toJSON();
    expect(tree).toMatchSnapshot();
})



const api = require('./../utils/api')
const {
    fetchData
} = api

//https://itnext.io/testing-components-with-jest-and-react-testing-library-d36f5262cde2
//https://kentcdodds.com/blog/fix-the-not-wrapped-in-act-warning
test('render when button click', async ()=> {

    const dataColor = require('../../public/color')

    const mockFn = jest.spyOn(window,'fetch').mockImplementationOnce(()=>{
        console.log('calling mock')
        return Promise.resolve({
            json: () => Promise.resolve(dataColor)
        })
    })

    const {container , findByRole} = render(<App/>)
    const button = container.querySelector('button')
    fireEvent.click(button)

    //findByRole interchangeble used with getByText and getByTestId
    const output = await waitForElement(()=> findByRole('output'))
    expect(mockFn).toHaveBeenCalled()
    expect(output).toBeDefined()

})



