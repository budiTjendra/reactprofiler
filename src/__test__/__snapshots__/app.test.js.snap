// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render correctly 1`] = `
<div
  className="App-fullscreen"
>
  <button
    data-testid="show-color-button"
    onClick={[Function]}
  >
    show color
  </button>
  <div
    style={
      Object {
        "height": 0,
        "overflow": "visible",
        "width": 0,
      }
    }
  />
</div>
`;
